packages/server/src/service-manager.ts

```diff
@@ -40,35 +40,38 @@ export class ServiceWorkerManager implements IServiceWorkerManager {
 
     if (!serviceWorker) {
       console.warn('ServiceWorkers not supported in this browser');
-    } else if (serviceWorker.controller) {
-      registration =
-        (await serviceWorker.getRegistration(serviceWorker.controller.scriptURL)) ||
-        null;
+      this._ready.reject('Not supported'); // Reject the promise if service workers are not supported.
+      return;
+    }
+
+    // Unregister any existing service workers before registering the new one.
+    const registrations = await serviceWorker.getRegistrations();
+    for (const registration of registrations) {
+      await registration.unregister();
```

Comment:
Curious, does that mean the service worker is unregistered and registered again at each page reload, not just a hard refresh?

This seems to be the case when testing on the preview: https://jupyterlite--1251.org.readthedocs.build/en/1251/_static/lab/index.html

![image](https://github.com/jupyterlite/jupyterlite/assets/591645/186c4105-b06c-4fa1-9e47-84c42ba2069f)


Not sure if that's the desired behavior, as noted in https://github.com/jupyterlite/jupyterlite/issues/751.

Reply:
Yeah I think you're right. Question is if we can detect hard refresh (that's when you have to unregister + register). I'll look into it.

I'm curious: that issue from 2022 (still open) indicates that we already do that, but I don't think we do?

Reply:
See the updated PR which is much simplified. Back to the old detection logic on existence, but added the `unregister` part before registering a new service worker. I'll verify also with the PR server that the reported bug is fixed with this change.

Reply:
> I'm curious: that issue from 2022 (still open) indicates that we already do that, but I don't think we do?

Indeed. From that issue there is also https://github.com/jupyterlite/jupyterlite/pull/871 which seems to have changed a bit the logic of the service worker.

Reply:
Are the tests flaky, or can you rerun? I'm not sure what fails with Firefox (I can successfully switch between lab and notebook in firefox myself)

Reply:
Yes the firefox UI test is still a bit flaky.

Reply:
Yeah, I see latest master is failing on this as well. Suppose I consider this PR ready then.